testing/integration/upgrade_standalone_inprogress.go

```diff
@@ -0,0 +1,97 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package integration
+
+import (
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	atesting "github.com/elastic/elastic-agent/pkg/testing"
+	"github.com/elastic/elastic-agent/pkg/testing/define"
+	"github.com/elastic/elastic-agent/testing/upgradetest"
+)
+
+// TestStandaloneUpgradeFailsWhenUpgradeIsInProgress initiates an upgrade for a
+// standalone Elastic Agent and, while that upgrade is still in progress, attempts
+// to initiate a second upgrade. The test expects Elastic Agent to not allow
+// the second upgrade.
+func TestStandaloneUpgradeFailsWhenUpgradeIsInProgress(t *testing.T) {
+	define.Require(t, define.Requirements{
+		Local: false, // requires Agent installation
+		Sudo:  true,  // requires Agent installation
+	})
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// For this test we start with a version of Agent that's two minors older
+	// than the current version and upgrade to the current version. Then we attempt
+	// upgrading to the current version again, expecting Elastic Agent to disallow
+	// this second upgrade.
+	upgradeFromVersion, err := upgradetest.PreviousMinor(ctx, define.Version())
+	require.NoError(t, err)
+	startFixture, err := atesting.NewFixture(
+		t,
+		upgradeFromVersion,
+		atesting.WithFetcher(atesting.ArtifactFetcher()),
+	)
+	require.NoError(t, err, "error creating previous agent fixture")
+
+	endFixture, err := define.NewFixture(t, define.Version())
+	require.NoError(t, err)
+	endVersionInfo, err := endFixture.ExecVersion(ctx)
+	require.NoError(t, err)
+
+	// Use the post-upgrade hook to bypass the remainder of the PerformUpgrade
+	// because we want to do our own checks for the rollback.
+	var ErrPostExit = errors.New("post exit")
+	postUpgradeHook := func() error {
+		return ErrPostExit
+	}
+
+	err = upgradetest.PerformUpgrade(
+		ctx, startFixture, endFixture, t,
+		upgradetest.WithPostUpgradeHook(postUpgradeHook))
+	if !errors.Is(err, ErrPostExit) {
+		require.NoError(t, err)
+	}
+
+	// upgrade has occurred and the watcher is running
+
+	// wait for the agent to be healthy and correct version
+	// before trying to perform another upgrade
+	err = upgradetest.WaitHealthyAndVersion(ctx, startFixture, endVersionInfo.Binary, 2*time.Minute, 10*time.Second, t)
+	if err != nil {
+		// agent never got healthy, but we need to ensure the watcher is stopped before continuing
```

Comment:
It wasn't immediately obvious to me _why_ the watcher was being killed here.  I think it's so it doesn't interfere with _other_ tests (as opposed to having anything to do with continuing on in this test, since this test should now fail due to the `require.NoError(t, err)` on line 79), right?  Could you clarify that in the code comment here please?

Reply:
I added more to the comment to make this clear. The reason you gave is the correct.